src/replayq.erl

```diff
@@ -226,13 +228,21 @@ append(#{config := #{seg_bytes := BytesLimit, dir := Dir} = Config,
 
 %% @doc pop out at least one item from the queue.
 %% volume limited by `bytes_limit' and `count_limit'.
--spec pop(q(), #{bytes_limit => bytes(), count_limit => count()}) ->
+-spec pop(q(),
+          #{
+            bytes_limit => bytes(),
+            count_limit => count(),
+            stop_before => stop_function(),
+            stop_before_input_accumulator => term()
```

Comment:
```suggestion
            accumulator => {accumulator_func(), accumulator_initial_state()}.
```
default accumulator:
```
%% in this module
-export([default_accumulator/2]).
default_accumulator(Item, List) -> [Item | List].
```
batch stopper accumulator:

```
%% in the caller module
-export([collector/2]).
collector(Item, []) -> [Item];
collector(Item, [Last | _] = List) ->
  case can_be_batched(Item, Last) of
    true ->
      [Item | List];
    false ->
      false
end.
```

Reply:
Thanks for the suggestion. Why do you think the function should be called accumulator/accumulator_func? Its purpose is not to accumulate but to decide when to stop adding items.

Reply:
I think I understand what you mean now. You want this to be more general so that one can can accumulate items however one wants (for example, into something else than a list), right? I think this complicates things in a way that does not benefit our current use case and it will break the current interface since the list needs to be reversed in the end by the user if one does not add a special case. If we want to have a custom accumulator function we can add that as another option. However, I think making the stop_before and stop_before_input_accumulator options a single option makes sense so I will use this suggestion.

Reply:
ok. can be a future enhancement as a `pop2` function.
this is beneficial when for example, we can encode a Kafka batch while popping, it gives more accuracy to the batch size.

currently the batch size is only an approximate estimation which can actually go over the limit due to the overhead bytes of each item, so we had to perform some oddly complicated compensation in estimation algorithms.